bindings/python/iree/compiler/debugging.py

```diff
@@ -0,0 +1,172 @@
+import threading
+
+_thread_locals = threading.local()
+_invocation_id = 0
+
+
+def _get_temp_file_saver_stack():
+  try:
+    return _thread_locals.temp_file_saver_stack
+  except AttributeError:
+    stack = []
+    _thread_locals.temp_file_saver_stack = stack
+    return stack
+
+
+def _interpolate_path_pattern(path_pattern: str, *, invocation_id: str):
+  # We do not use str.format() because we do not know the providence of
+  # path_pattern. Instead, handle a fixed set of replacements.
+  path_pattern = path_pattern.replace("{id}", str(invocation_id))
+  path_pattern = path_pattern.replace("{pid}", str(os.getpid()))
+  path_pattern = path_pattern.replace("{main}", os.path.basename(sys.argv[0]))
+  return path_pattern
+
+
+class TempFileSaver:
+  """Manages the saving of temp files resulting from tool invocations.
+
+  The TempFileSaver is a thread-local context bound object. An attempt to
+  create a new one will return the most recent instance created and entered
+  as a context manager. This allows up-stack callers to establish the
+  policy for saving temporaries and deep implementations will inherit it.
+
+  Proper usage from users wishing to establish a saver context:
+    with TempFileSaver():
+      # Do things with temp files.
+
+  Proper usage for implementors wishing to use an established saver context
+  or set up a new one:
+    with TempFileSaver.implicit() as tfs:
+      # Do things with temp files.
+
+  The outer-most creator can customize it with explicit arguments to __init__
+  but these will be ignored if an instance is already thread bound.
+  """
+  TEMP_PATH_ENV_KEY = "IREE_SAVE_TEMPS"
+
+  @staticmethod
+  def implicit():
+    stack = _get_temp_file_saver_stack()
+    if stack:
+      return stack[-1]
+    return TempFileSaver()
+
+  def __init__(self,
+               temp_path_pattern: str = None,
+               *,
+               invocation_id: Optional[str] = None):
+    self.retained = False
+    self._refcount = 0
+    if temp_path_pattern is None:
+      temp_path_pattern = os.environ.get(TempFileSaver.TEMP_PATH_ENV_KEY)
+    if temp_path_pattern is None:
+      return
+
+    global _invocation_id
+    if invocation_id is not None:
+      self.invocation_id = invocation_id
+    else:
+      self.invocation_id = _invocation_id
+      _invocation_id += 1
+
+    self.retained_path = _interpolate_path_pattern(
+        temp_path_pattern, invocation_id=self.invocation_id)
+    self.retained = True
+    self._retained_file_names = set()
+    self._copy_on_finalize = list()  # Of (source_path, target_path)
+    self._created_dir = False
+
+  def __enter__(self):
+    _get_temp_file_saver_stack().append(self)
+    self._refcount += 1
+    return self
+
+  def __exit__(self, exc_type, exc_value, traceback):
+    del _get_temp_file_saver_stack()[-1]
+    self._refcount -= 1
+    if self._refcount == 0:
+      self._finalize()
+
+  @staticmethod
+  def current():
+    try:
+      return _get_temp_file_saver_stack()[0]
```

Comment:
Shouldn't this be the last entry on the stack?

Reply:
indeed. thanks.